src/Pipeline/Queue/Atomic.php

```diff
@@ -0,0 +1,215 @@
+
+            return;
+        }
+
+        $this->txCommands->enqueue($command);
+        $this->innerQueue->enqueue($command);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function flush(ConnectionInterface $connection): Traversable
+    {
+        if ($this->txCommands->isEmpty()) {
+            yield from [];
+
+            return 0;
+        }
+
+        $this->innerQueue->enqueue(RawCommand::create('EXEC'));
+
+        $pending = count($this->txCommands);
+        $responses = new NoRewindIterator($this->innerQueue->flush($connection));
+
+        if (!$responses->valid()) {
+            yield from [];
+
+            return 0;
+        }
+
+        // NOTE: first queued command is `MULTI` so we check its response as it
+        // could return "-ERR" if Redis detects a nested `MULTI` when the atomic
+        // queue is flushed over a connection in a running MULTI / EXEC context.
+        if ($responses->current() instanceof ErrorInterface) {
+            throw $this->onLogicError($connection, $responses->current()->getMessage());
+        }
+
+        for ($i = 0; $i <= $pending; $i++) {
+            $responses->next();
+        }
+
+        /** @var iterable|ErrorInterface */
+        $execResponse = $responses->current();
+
+        $this->ensureValidExecResponse($connection, $execResponse);
+
+        foreach ($execResponse as $response) {
+            yield $this->txCommands->dequeue() => $response;
+        }
+
+        return count($execResponse);
+    }
+
+    /**
+     * Makes sure the response payload returned by `EXEC` is valid.
+     *
+     * @param ConnectionInterface     $connection   Target connection
+     * @param iterable|ErrorInterface $execResponse Response returned by `EXEC`
+     *
+     * @throws AtomicFlushException when the response returned by `EXEC` is not valid
+     */
+    protected function ensureValidExecResponse(ConnectionInterface $connection, $execResponse): void
+    {
+        if (null === $execResponse) {
+            $this->reset();
+
+            throw $this->onAbortedError($connection, 'Transaction discarded because of previous errors (NULL response)');
+        } elseif ($execResponse instanceof ErrorInterface) {
+            $this->reset();
+
+            if ('EXECABORT' === $execResponse->getErrorType()) {
+                throw $this->onAbortedError($connection, 'Transaction discarded because of previous errors (-EXECABORT response)');
+            } else {
+                throw $this->onLogicError($connection, $execResponse->getMessage());
+            }
+        } elseif (!is_iterable($execResponse)) {
+            $this->reset();
+            $connection->disconnect();
+
+            throw $this->onStateError($connection, sprintf(
+                'Protocol desynchronization detected on `EXEC` response (array expected, `%s` received)',
+                is_object($execResponse) ? get_class($execResponse) : gettype($execResponse)
+            ));
+        }
```

Comment:
This is the only section I'll need to take some time out and take a proper look at later, but I've gone through everything else and it looks good.

Good job @nrk

Reply:
Let me know if you need any clarification.

Reply:
Will do 😎